### DISCUSSION TITLE
[Draft proposal] : auto namespace (Define namespace depending on location file)

### DISCUSSION THREAD

**Discussion by ToovR** [dotnet/csharplang#7908] [Language Ideas]
# auto namespace
- [X] Proposed
- [ ] Prototype
- [ ] Implementation: Not Started
- [ ] Specification: Not started

## Summary
auto namespace is a feature which avoids writing namespace name and evaluates namespace name based on project namespace and file location.

## Motivation
When you need to move a C# file in your project or from a project to another, you need to change your namespace name definition to fit the new project, new folder location.

With auto namespace, you do not need to care about naming the namespace. It would be automatically defined on project building depending the namespace of the project and the file location.

Auto namespace could also avoid Typos when manually declare a namespace.

For example, let's imagine a C# class library project "Some.Namespace.MyLibraryClass" with a class file stored in the project as "./Folder/Stored/".
We should have by default one of the following namespace declarations :

```CSharp
namespace Some.Namespace.MyLibraryClass.Folder.Stored;
```

```CSharp
namespace Some.Namespace.MyLibraryClass.Folder.Stored
{

}
```

by replacing with a specific key word, to be defined but here written as "auto namespace", we could write as follows and simplify moving files, and avoid typos:

```CSharp
auto namespace; 
```

```CSharp
auto namespace
{

}
```

**Comment by huoyaoyuan** [comment]
#2908

> I'm not a fan of having code files have to be aware of the file system.

**Comment by colejohnson66** [comment]
> When you need to move a C# file in your project or from a project to another, you need to change your namespace name definition to fit the new project, new folder location.

IMO, any competent IDE should offer to update namespaces for you, either on file move or with a warning on the namespace. Rider, for example, does.
